### PR TITLE
Proposed fix for issue #4443

### PR DESCRIPTION
**Pull request by juan42**
fixes https://github.com/dealii/dealii/issues/4443

### CODE REVIEW

**Comment by davydden** [comment]
please add a changelog entry in `doc/news/changes/minor`.

File: include/deal.II/lac/trilinos_solver.h [hunk 1]
```diff
@@ -364,24 +358,8 @@ namespace TrilinosWrappers
...
-      explicit
-      AdditionalData (const bool output_solver_details = false);
-
-      /**
-       * Enables/disables the output of solver details (residual in each
-       * iterations etc.).
-       */
-      bool output_solver_details;
-    };
+    using SolverBase::AdditionalData;
```
Comment (jppelteret) [inline include/deal.II/lac/trilinos_solver.h]: You need to add this line to all of the other solver classes as well. It would also be good to keep the documentation.

File: include/deal.II/lac/trilinos_solver.h [hunk 2]
```diff
@@ -114,25 +114,19 @@ namespace TrilinosWrappers
        * base class for all solvers.
        */
       explicit
-      AdditionalData (const bool         output_solver_details   = false,
-                      const unsigned int gmres_restart_parameter = 30);
```
Comment (jppelteret) [inline include/deal.II/lac/trilinos_solver.h]: Why did you need to remove the GMRES parameter?

File: source/lac/trilinos_solver.cc [hunk 3]
```diff
@@ -449,7 +448,7 @@ namespace TrilinosWrappers
         break;
       case gmres:
         solver.SetAztecOption(AZ_solver, AZ_gmres);
-        solver.SetAztecOption(AZ_kspace, additional_data.gmres_restart_parameter);
+        // solver.SetAztecOption(AZ_kspace, additional_data.gmres_restart_parameter);
```
Comment (jppelteret) [inline source/lac/trilinos_solver.cc]: Again, can you briefly explain why you needed to remove this?

File: include/deal.II/lac/trilinos_solver.h (hunk 2 above)
Comment (juan42) [inline include/deal.II/lac/trilinos_solver.h]: Otherwise, every inheritance from SolverBase would need to have this parameter. For SolverCG this doesn't make sense.

Comment (jppelteret) [inline include/deal.II/lac/trilinos_solver.h]: But doing this breaks [this functionality](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/trilinos_solver.h#L72) of using `SolverBase` as a "SolverSelector". I can imagine that this would cause both our unit tests and some user codes to fail.

As far as I can see, only other way to implement the necessary fix would be to revert to the original structure and to do the following:
```c++
  SolverCG::SolverCG (SolverControl        &cn,
                      const AdditionalData &data) // This is SolverCG's AdditionalData
    :
    SolverBase (cn,
                SolverBase::AdditionalData(data.output_solver_details)), // Pass down flags to base class. Its gmres_restart_parameter would take on a default value.
    additional_data (data.output_solver_details)
  {
    solver_name = cg;
  }
```
As discussed in our original conversation, I would consider this a suitable fix as well (its actually the "easier" of the two options).

Comment (juan42) [inline include/deal.II/lac/trilinos_solver.h]: And then you would again get the problem, that SolverBase doesn't use that object in the do_solve routine where it needs the options. At least i have tried it and couldn't find a way getting that to work. I also get two structs of AdditionalData (one from the inherited and one from SolverBase) in the SolverCG when i inspect the object in a debugger. Maybe i don't understand something here.

Comment (jppelteret) [inline include/deal.II/lac/trilinos_solver.h]: Do you have a small test case that you're using to check the output of the solver? If so, could you share it and I'll take a look at this afternoon. I'm clearly not understanding why there's an issue, so it would be better if I just get my hands dirty and fiddle with it as well :-)

**Comment by masterleinad** [comment]
@jppelteret @juan42 This PR seems to superseded by #4515. Can it be closed?
